sound/soc/sof/control.c

```diff
@@ -342,24 +345,55 @@ int snd_sof_bytes_ext_get(struct snd_kcontrol *kcontrol,
 	struct snd_sof_dev *sdev = scontrol->sdev;
 	struct sof_ipc_ctrl_data *cdata = scontrol->control_data;
 	unsigned int tag = 0;
+	int max_length = SOF_IPC_MSG_MAX_SIZE -
+		sizeof(const struct sof_ipc_ctrl_data) -
+		sizeof(const struct sof_abi_hdr);
+	int err;
+	int ret;
 
-	pm_runtime_get_sync(sdev->dev);
+	ret = pm_runtime_get_sync(sdev->dev);
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: bytes get failed to resume %d\n",
+			ret);
+		return ret;
+	}
 
-	dev_dbg(sdev->dev, "getting data and command is %d\n", scontrol->cmd);
-	/* get all the mixer data from DSP */
+	/* Decrement size for room for ext bytes header and find out
+	 * the limit from ext bytes size from topology and SOF IPC
+	 * size limit.
+	 */
+	size -= 2 * sizeof(u32);
+	if (max_length > be->max)
+		max_length = be->max;
+
+	dev_dbg(sdev->dev, "request size minus header is %d\n", size);
+	dev_dbg(sdev->dev, "max size is %d\n", max_length);
+
+	/* Check size to prevent read of other kernel data */
+	if (size > max_length) {
+		dev_err(sdev->dev, "error: size is too large\n");
+		ret = -EINVAL;
+		goto out;
+	}
+
+	/* get all the component data from DSP */
 	snd_sof_ipc_get_comp_data(sdev->ipc, scontrol, SOF_IPC_COMP_GET_DATA,
 				  SOF_CTRL_TYPE_DATA_GET, scontrol->cmd);
 
-	/* TODO: replace 252 with actual size */
 	if (copy_to_user(binary_data, &tag, sizeof(u32)))
 		return -EFAULT;
+
 	if (copy_to_user(binary_data + 1, &size, sizeof(u32)))
 		return -EFAULT;
-	if (copy_to_user(binary_data + 2, cdata->data->data,
-			 SOF_IPC_MSG_MAX_SIZE - sizeof(struct sof_ipc_ctrl_data)))
+
+	if (copy_to_user(binary_data + 2, cdata->data->data, size))
 		return -EFAULT;
 
+out:
 	pm_runtime_mark_last_busy(sdev->dev);
-	pm_runtime_put_autosuspend(sdev->dev);
-	return 0;
+	err = pm_runtime_put_autosuspend(sdev->dev);
+	if (err < 0)
+		dev_err(sdev->dev, "error: failed to idle %d\n", err);
+
+	return ret;
```

Comment:
@singalsu I see why you have err and ret, but its confusing to see err just above and return "ret" instead. I think we should just use one and return the latest return value from the runtime_put. What do you think?

Reply:
That's a good question. I need to find out if err means that the operation has failed with incorrect data or is this just something to note into debug trace. Failing with idle doesn't sound like the ext bytes get has failed. I never saw this to happen in my tests though.

There's also the return value from the IPC command that is not stored & checked. That might be the most relevant for return code.

Reply:
@singalsu I think we should skip checking the return value for runtime_put if we arent going to use the value at all.
It is more important to check the return value for runtime_get_sync.